File: app/androidApp/app/src/main/java/com/lynxapp/LynxGenericResourceFetcher.kt
// Copyright 2024 The Lynx Authors. All rights reserved.
// Licensed under the Apache License Version 2.0 that can be found in the
// LICENSES/Apache-2.0.txt file in the root directory of this repository.
// Modified by the Lynx template project to use its shared application transport.

package com.lynxapp

import com.lynx.tasm.resourceprovider.LynxResourceCallback
import com.lynx.tasm.resourceprovider.LynxResourceRequest
import com.lynx.tasm.resourceprovider.LynxResourceResponse
import com.lynx.tasm.resourceprovider.generic.LynxGenericResourceFetcher
import java.io.IOException
import okhttp3.Call
import okhttp3.Callback
import okhttp3.Request
import okhttp3.Response

/**
 * App-owned fetcher for runtime-loaded Lynx resources such as rspeedy HMR
 * patches, which the engine requests outside the template provider flow.
 */
internal object LynxGenericResourceFetcher : LynxGenericResourceFetcher() {
    override fun fetchResource(
        request: LynxResourceRequest,
        callback: LynxResourceCallback<ByteArray>,
    ) {
        val call = runCatching {
            val url = Request.Builder().url(request.url).build().url
            require(BuildConfig.DEBUG || url.isHttps) {
                "Cleartext HTTP is disabled in release builds"
            }
            AppHttpClient.instance.newCall(Request.Builder().url(url).build())
        }.getOrElse { error ->
            deliverFailure(request.url, error, callback)
            return
        }

        call.enqueue(object : Callback {
            override fun onFailure(call: Call, e: IOException) {
                deliverFailure(request.url, e, callback)
            }

            override fun onResponse(call: Call, response: Response) {
                response.use {
                    val body = runCatching { it.body?.bytes() }.getOrElse { error ->
                        deliverFailure(request.url, error, callback)
                        return
                    }
                    if (!it.isSuccessful) {
                        deliverFailure(
                            request.url,
                            IOException("HTTP ${it.code} for ${request.url}"),
                            callback,
                        )
                        return
                    }
                    callback.onResponse(LynxResourceResponse.onSuccess(body ?: ByteArray(0)))
                }
            }
        })
    }

    override fun fetchResourcePath(
        request: LynxResourceRequest,
        callback: LynxResourceCallback<String>,
    ) {
        deliverFailure(request.url, UnsupportedOperationException("unsupported: ${request.url}"), callback)
    }

    override fun cancel(request: LynxResourceRequest) {
        // The shared OkHttp client does not expose per-request cancellation;
        // in-flight reads complete and are dropped when the view is destroyed.
    }

    @Suppress("UNCHECKED_CAST")
    private fun <T> deliverFailure(
        url: String,
        error: Throwable,
        callback: LynxResourceCallback<T>,
    ) {
        if (BuildConfig.DEBUG) android.util.Log.w(TAG, "resource fetch failed: $url", error)
        callback.onResponse(LynxResourceResponse.onFailed(error) as LynxResourceResponse<T>)
    }

    private const val TAG = "LynxResourceFetch"
}

File: app/iosApp/iosApp/LynxGenericResourceFetcher.m
// Copyright 2024 The Lynx Authors. All rights reserved.
// Licensed under the Apache License Version 2.0 that can be found in the
// LICENSES/Apache-2.0.txt file in the root directory of this repository.
// Implemented for the Lynx template project to load runtime-requested Lynx
// resources such as rspeedy HMR patches outside the template provider flow.

#import "LynxGenericResourceFetcher.h"

#import <Lynx/LynxResourceRequest.h>

@implementation AppGenericResourceFetcher

- (dispatch_block_t)fetchResource:(LynxResourceRequest *)request
                        onComplete:(LynxGenericResourceCompletionBlock)callback {
  NSURL *url = [NSURL URLWithString:request.url];
  NSString *scheme = url.scheme.lowercaseString;
  BOOL cleartextAllowed = NO;
#if DEBUG
  cleartextAllowed = YES;
#endif
  if (url == nil || scheme.length == 0) {
    callback(nil, [self errorWithMessage:[NSString
        stringWithFormat:@"Invalid resource URL: %@", request.url]]);
    return ^{};
  }
  if (![scheme isEqualToString:@"https"] &&
      (!cleartextAllowed || ![scheme isEqualToString:@"http"])) {
    callback(nil, [self errorWithMessage:[NSString
        stringWithFormat:@"Cleartext HTTP is disabled in release builds: %@",
                         request.url]]);
    return ^{};
  }

  NSURLSessionDataTask *task =
      [[NSURLSession sharedSession] dataTaskWithURL:url
        completionHandler:^(NSData *_Nullable data, NSURLResponse *_Nullable response,
                            NSError *_Nullable error) {
          if (error != nil) {
            callback(nil, error);
            return;
          }
          if (![response isKindOfClass:[NSHTTPURLResponse class]]) {
            callback(nil, [self errorWithMessage:@"Resource response was not HTTP"]);
            return;
          }
          NSInteger status = ((NSHTTPURLResponse *)response).statusCode;
          if (status < 200 || status >= 300) {
            callback(nil, [self errorWithMessage:[NSString
                stringWithFormat:@"HTTP %ld for %@",
                                 (long)status, url.absoluteString]]);
            return;
          }
          callback(data ?: [NSData data], nil);
        }];
  [task resume];
  return ^{ [task cancel]; };
}

- (dispatch_block_t)fetchResourcePath:(LynxResourceRequest *)request
                           onComplete:(LynxGenericResourcePathCompletionBlock)callback {
  callback(nil, [self errorWithMessage:[NSString
      stringWithFormat:@"fetchResourcePath is unsupported for %@", request.url]]);
  return ^{};
}

- (NSError *)errorWithMessage:(NSString *)message {
  return [NSError errorWithDomain:@"com.lynxapp.generic-resource-fetcher"
                             code:-1
                         userInfo:@{NSLocalizedDescriptionKey : message}];
}

@end
